Initial implementation

jszwedko opened the pull request.

jszwedko committed:
Initial implementation

alexschlessinger commented:
lgtm :+1:

rafecolton commented on README.md:
````diff
@@ -0,0 +1,76 @@
+
+```yml
+confd_version: # version of confd to download
+confd: # confd remote filename
````
Some of these variable names are unclear.  For example, this one should probably be `confd_remote_filename`, since that's what it is

jszwedko replied:
:+1:

rafecolton commented on defaults/main.yml:
```diff
@@ -0,0 +1,10 @@
+---
+confd_version: 0.6.3
+confd: "confd-{{ confd_version }}-linux-amd64"
+confd_download: "https://github.com/kelseyhightower/confd/releases/download/v{{ confd_version }}/{{ confd }}"
```
You can probably condense `confd`, `confd_download`, and `confd_vesrion` into `confd_download_url`

jszwedko replied:
I'd like to leave the version separate, but I'm OK with dropping `confd` and assuming `-linux-amd64`

rafecolton replied:
Honestly having them all separate reminds me of the way we used to do attributes in chef - having way more vars than we need.

rafecolton commented on README.md:
```diff
@@ -0,0 +1,76 @@
+confd_version: # version of confd to download
+confd: # confd remote filename
+confd_download: # url to download confd from
+confd_home: # where to put the confd binary
```
This is also a bit unclear - might be better to rename it and to have it be the full path to the downloaded binary (i.e. including the binary name)

jszwedko replied:
:+1:, I'll use `confd_binary_path`.

rafecolton commented on README.md:
```diff
@@ -0,0 +1,76 @@
+confd: # confd remote filename
+confd_download: # url to download confd from
+confd_home: # where to put the confd binary
+confd_config: # where to store confd configuration
```
This should probably be `confd_config_directory` so it is clear that the var should not contain confd configuration data

jszwedko replied:
:+1:

rafecolton commented on tasks/main.yml:
```diff
@@ -0,0 +1,33 @@
+    dest="{{confd_config}}/templates/{{item.template_name|default(item.service_name + '.conf.tmpl')}}"
+  with_items: confd_resources
+
+- name: add cron task
```
This should probably be an upstart service and pass `-interval 60` instead of being a cron task

jszwedko replied:
What advantages would that offer?

rafecolton replied:
The system can manage it as a service and we can monitor the health of the service instead of determining its success or failure from logs

jszwedko replied:
Gotcha, you've persuaded me.

rafecolton commented on templates/resource.toml.j2:
```diff
@@ -0,0 +1,16 @@
+  {% endfor %}
+]
+check_cmd = "{{item.check_cmd | default('true')}}"
+reload_cmd = "{{item.reload_cmd | default('sudo service ' + item.service_name + ' restart')}}"
```
This should probably default to not having a reload command at all.  Also, it shouldn't use `service`, it should use the upstart command instead (`initctl`)

jszwedko replied:
It's just a default that works on Ubuntu Trusty, but I agree that the default should be to leave the key out.

rafecolton commented on templates/resource.toml.j2:
```diff
@@ -0,0 +1,16 @@
+    "{{key}}",
+  {% endfor %}
+]
+check_cmd = "{{item.check_cmd | default('true')}}"
```
Same here - this should probably default to the key not being present if it is not defined

rafecolton commented on README.md:
```diff
@@ -0,0 +1,76 @@
+# these are used to generate confd resource.toml's and templates to use during configuration generation
+confd_resources: []
+# each resource is a hash with the following keys:
+# service_name (required): name of the service which uses the confd generated configuration
```
We should probably make it clear that this should have no spaces in it, given how it is used in the templates

jszwedko replied:
:+1:

rafecolton commented on templates/resource.toml.j2:
```diff
@@ -0,0 +1,16 @@
+
+[template]
+src = "{{item.template_name | default(item.service_name + '.conf.tmpl')}}"
+dest = "{{item.template_output | default('/etc/default/' + item.service_name)}}"
```
This seems like an odd default - very specific to one use case.  I would rather `template_output` be required and not defaulted

jszwedko replied:
Agree :+1:

rafecolton commented on tasks/main.yml:
```diff
@@ -0,0 +1,33 @@
+---
```
Might be good to run `confd -onetime` (with whatever other args) at the end of the tasks as a way of validating the templates and keys

jszwedko replied:
:+1:

jszwedko committed:
Rename variables to be more descriptive

jszwedko committed:
Making all resource variables optional

Only writing them if they are set.

Also renamed some to be consistent.

jszwedko committed:
Run confd once during playbook run to validate the configuration

jszwedko commented:
@rafecolton I think I addressed all of your comments

rafecolton commented:
Overall looks good.  I can live with the run being a cron task, objection to that isn't super strong.  I would also love to have some tests, but that can be a subsequent pull request.  Let me know if you want me to merge this :+1:

jszwedko committed:
Add whitespace

jszwedko committed:
Using upstart rather than cronjob to run confd

So the status can be monitored

jszwedko commented:
@rafecolton too late! Added upstart task.

jszwedko committed:
Fixing whitespace in upstart template

jszwedko committed:
Make sure service is started (in case the configuration didn't change)

jszwedko committed:
Re-add running of confd during play

To ensure configuration is valid.

jszwedko committed:
Revising variable names and optionalness

Also replacing %s with {{.src}} in the check_cmd to get around Jinja
templating. Couldn't find any other way that didn't result in {#.src#}.

jszwedko committed:
Add tests

jszwedko committed:
confd_binary_location -> confd_binary_path

To match tasks/default.yml

jszwedko committed:
Use 'env' as backend on travis

Easier than installing etcd or consul

jszwedko committed:
etcd_backend -> confd_backend

Typo
